### PR TITLE
crimson/os/seastore/circular_bounded_journal: fixes and cleanups

### PR DESCRIPTION
**Pull request by cyx1231st**
* replace config.start to CBJOURNAL_START_ADDRESS.
* do not split records due to relative paddr resolution.
* only increment seq when roll.
* cleanups.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by athanatos — APPROVED** [review]
Seems ok to me with two notes.

I think @myoungwon had a purpose in mind for applied_to, but I'm happy to remove it until we add a user.

Are we avoiding splitting records at the end of the journal mainly to simplify the implementation?

File: src/crimson/os/seastore/journal/circular_bounded_journal.cc [hunk 1]
```diff
@@ -163,122 +158,69 @@ CircularBoundedJournal::open_device_read_header(rbm_abs_addr start)
...
+    if (encoded_size > get_available_size()) {
+      ERROR("rolled, record size {}, but available size {}",
+            encoded_size, get_available_size());
+      return crimson::ct_error::erange::make();
+    }
+  }
 
   journal_seq_t j_seq {
-    cur_segment_seq++,
+    circulation_seq,
```
Comment (myoungwon) [inline src/crimson/os/seastore/journal/circular_bounded_journal.cc:187]: Could you please explain the purpose of circulation_seq? Sequence number here is originally used to check that records are stored in order correctly---it was increased every record.

File: src/test/crimson/seastore/test_cbjournal.cc [hunk 2]
```diff
@@ -138,10 +138,9 @@ struct cbjournal_test_t : public seastar_test_suite_t
       epm(new ExtentPlacementManager(true)),
       cache(*epm)
   {
-    device = new nvme_device::TestMemory(CBTEST_DEFAULT_TEST_SIZE);
+    device = new nvme_device::TestMemory(CBTEST_DEFAULT_TEST_SIZE + CBTEST_DEFAULT_BLOCK_SIZE);
```
Comment (myoungwon) [inline src/test/crimson/seastore/test_cbjournal.cc:141]: Why do we need additional block size?

File: src/crimson/os/seastore/journal/circular_bounded_journal.cc (hunk 1 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/journal/circular_bounded_journal.cc:187]: In the segmented-journal, segment-seq only increase during rolling to a new segment, which is very similar to rolling to the start of the circular journal, it won't increment for every record, see https://github.com/ceph/ceph/blob/main/src/crimson/os/seastore/journal/segment_allocator.cc#L38-L39

Instead, `journal_seq_t` is the actual thing that will be increased by every record as it consists of [segment-seq, offset].

File: src/test/crimson/seastore/test_cbjournal.cc (hunk 2 above)
Comment (myoungwon) [reply src/test/crimson/seastore/test_cbjournal.cc:141]: We keep track of information about available size. Isn't it enough to check whether the journal is empty of full?

Comment (myoungwon) [reply src/test/crimson/seastore/test_cbjournal.cc:141]: I forgot that I removed available size variable in the header. But, I am not sure that using additional one block space here is better way. Can we add a variable in the header or in-memory variable in CBjournal? This does not require a physical block. What do you think?

Comment (myoungwon) [reply src/test/crimson/seastore/test_cbjournal.cc:141]: I got it.

**Review by myoungwon — APPROVED** [review]

**Pull request merged**